### PR TITLE
[Arc] RemoveUnusedArcArguments pass improvements

### PR DESCRIPTION
**Pull request by maerhart**
Most importantly, this adds support for the new `arc.call` operation. Otherwise, it will not consider it and thus lead to errors because the argument list does not match anymore after the pass.

Since removing arguments of one arc can now lead to more unused arguments in another arc due to the removal of operands in the `arc.call`, it is changed to use the pattern rewriter to check for these cases again instead of manually keeping track of it.
I also added some pass statistics to see how many arguments are actually removed in how many arcs.

The `CallOpMutableInterface` allows for easy generalization and addition of new operations that can act as call ops. I couldn't find an equivalent in MLIR. I don't know if it was a deliberate choice there to not have such an interface method, but otherwise it could be something to upstream.

### CODE REVIEW

File: include/circt/Dialect/Arc/ArcInterfaces.td [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  let methods = [
+    InterfaceMethod<[{
+        Returns the operands within this call that are used as arguments to the
+        callee as a mutable range.
+      }],
+      "::mlir::MutableOperandRange", "getArgOperandsMutable"
+    >,
+  ];
+}
```
Comment (fabianschuiki) [inline include/circt/Dialect/Arc/ArcInterfaces.td:29]: That's a very handy utility! I could totally see this being useful upstream as well.

**Review by fabianschuiki — APPROVED** [review]
Nice improvement, LGTM!

I wonder if we could do something similar with results… probably not as neatly because they can't be mutated as easily as operands if I recall correctly.

**Comment by maerhart** [comment]
> I wonder if we could do something similar with results… probably not as neatly because they can't be mutated as easily as operands if I recall correctly.

Great idea! It would indeed be a more complex rewrite pattern, but probably worth adding. Just need to update the pass name as well then 😄

**Comment by fabianschuiki** [comment]
> Just need to update the pass name as well then

Haha yeah. Sounds like an orthogonal improvement for the future, not today 😁

**Pull request merged**